lib/events/api.go

```diff
@@ -799,13 +799,28 @@ type UploadMetadataGetter interface {
 	GetUploadMetadata(sid session.ID) UploadMetadata
 }
 
-// StreamWriter implements io.Writer to be plugged into the multi-writer
-// associated with every session. It forwards session stream to the audit log
-type StreamWriter interface {
+// SessionEventSetter will set necessary event fields for session-related
+// events and must be called before the event is recorded, regardless
+// of whether the event will be recorded, emitted, or both.
+type SessionEventSetter interface {
+	SetupSessionEvent(event apievents.AuditEvent) (apievents.AuditEvent, error)
+}
+
+// SessionRecorder records session events. It can be used both as a
+// [io.Writer] when recording raw session data and as a [apievents.Recorder]
+// when recording session events.
+type SessionRecorder interface {
 	io.Writer
 	apievents.Stream
 }
 
+// SessionSetterRecorder sets necessary session event fields and records them.
+type SessionSetterRecorder interface {
+	SetupAndRecordEvent(ctx context.Context, event apievents.AuditEvent) error
```

Comment:
If a concrete type implements both `SessionEventSetter` and `SessionRecorder` doesn't it have everything it needs?

(Why do we need to require that it also implements `SetupAndRecordEvent`?)

Reply:
This was a bit of a smell, making `SetupAndRecordEvent` a function is a much better idea